opt_einsum/path_random.py

```diff
@@ -0,0 +1,320 @@
+    implemented with the following signature::
+
+        def setup(self, inputs, output, size_dict):
+            # custom preparation here ...
+            return trial_fn, trial_args
+
+    Where ``trial_fn`` itself should have the signature::
+
+        def trial_fn(r, *trial_args):
+            # custom computation of path here
+            return ssa_path, cost, size
+
+    Where ``r`` is the run number and could for example be used to seed a
+    random number generator. See ``RandomGreedy`` for an example.
+
+
+    Parameters
+    ----------
+    max_repeats : int, optional
+        The maximum number of repeat trials to have.
+    max_time : float, optional
+        The maximum amount of time to run the algorithm for.
+    minimize : {'flops', 'size'}, optional
+        Whether to favour paths that minimize the total estimated flop-count or
+        the size of the largest intermediate created.
+    executor : executor-pool like, optional
+        An executor-pool to optionally parallelize repeat trials over. The pool
+        should have an api matching those found in the python 3 standard library
+        module ``concurrent.futures``. Namely, a ``submit`` method that returns
+        ``Future`` objects, themselves with ``result`` and ``cancel`` methods.
+    pre_dispatch : int, optional
+        If using an ``executor``, how many jobs to pre-dispatch so as to avoid
+        submitting all jobs at once. Should also be more than twice the number
+        of workers to avoid under-subscription. Default: 128.
+
+    Attributes
+    ----------
+    path : list[tuple[int]]
+        The best path found so far.
+    costs : list[int]
+        The list of each trial's costs found so far.
+    sizes : list[int]
+        The list of each trial's largest intermediate size so far.
+
+    See Also
+    --------
+    RandomGreedy
+    """
+
+    def __init__(self, max_repeats=32, max_time=None, minimize='flops', executor=None, pre_dispatch=128):
+
+        if minimize not in ('flops', 'size'):
+            raise ValueError("`minimize` should be one of {'flops', 'size'}.")
+
+        self.max_repeats = max_repeats
+        self.max_time = max_time
+        self.minimize = minimize
+        self.better = _BETTER_FNS[minimize]
+        self.executor = executor
+        self.pre_dispatch = pre_dispatch
+
+        self.costs = []
+        self.sizes = []
+        self.best = {'flops': float('inf'), 'size': float('inf')}
+
+    @property
+    def path(self):
+        """The best path found so far.
+        """
+        return ssa_to_linear(self.best['ssa_path'])
+
+    def _gen_results_executor(self, repeats, trial_fn, args):
+        """Lazily generate results from an executor without submitting all jobs at once.
+        """
+        self._futures = deque()
+
+        for r in repeats:
+            if len(self._futures) < self.pre_dispatch:
+                self._futures.append(self.executor.submit(trial_fn, r, *args))
+                continue
```

Comment:
Do you need this continue?

Reply:
So we need the ``continue`` and the second ``yield`` below in conjunction - the continue means that ``pre_dispatch`` number of jobs are submitted before we start yielding futures (otherwise only 1 would). The second yield is to then exhaust the queue once we have submitted all the jobs.

Reply:
Ah, gotcha. Makes sense, could you add a comment or two explaining this.